prefetchLanguages.nix

```diff
@@ -0,0 +1,114 @@
+{ pkgs, ... }:
+
+let
+  inherit (builtins)
+    attrNames
+    concatStringsSep
+    mapAttrs
+    toFile
+    readFile
+    toJSON
+    fromJSON
+    baseNameOf
+    ;
+
+  inherit (pkgs) fetchgit nickel;
+  inherit (pkgs.lib) warn;
+  inherit (pkgs.stdenvNoCC) mkDerivation;
+  inherit (pkgs.tree-sitter) buildGrammar;
+
+  prefetchLanguageSourceGit =
+    name: source:
+    buildGrammar {
+      language = name;
+      version = source.rev;
+      src = fetchgit {
+        url = source.git;
+        rev = source.rev;
+        hash =
+          if source ? "nixHash" then
+            source.nixHash
+          else
+            warn "Language ${name}: no nixHash provided - using dummy value" "sha256-AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA=";
+      };
+      location = if source ? "subdir" then source.subdir else null;
+    };
+
+  prefetchLanguageSource =
+    name: source:
+    if source ? "path" then
+      { inherit (source) path; }
+    else if source ? "git" then
+      { path = "${prefetchLanguageSourceGit name source.git}/parser"; }
+    else
+      throw ("Unsupported Topiary language sources: " ++ concatStringsSep ", " (attrNames source));
+
+  ## Given a Topiary configuration as a Nix value, returns the same
+  ## configuration, except all language sources have been replaced by a
+  ## prefetched and precompiled one. This requires the presence of a `nixHash`
+  ## for all sources.
+  prefetchLanguages =
+    topiaryConfig:
+    topiaryConfig
+    // {
+      languages = mapAttrs (
+        name: languageConfig:
+        languageConfig
+        // {
+          grammar = languageConfig.grammar // {
+            source = prefetchLanguageSource name languageConfig.grammar.source;
+          };
+        }
+      ) topiaryConfig.languages;
+    };
+
+  toJSONFile =
+    name: e:
+    mkDerivation {
+      name = "${name}.json";
+      dontUnpack = true;
+      buildPhase = ''
+        cat <<\EOF > $out
+        ${toJSON e}
+        EOF
+      '';
+    };
+
+  toNickelFile =
+    name: e:
+    mkDerivation {
+      name = "${name}.ncl";
+      dontUnpack = true;
+      buildPhase = ''
+        echo "import \"${toJSONFile name e}\"" > $out
```

Comment:
We also sometimes use `nickel eval $file` to produce a Nickel translation of a JSON file (since `eval` prints the result as a Nickel value). But it's not really meant for this, so until we have a proper `nickel convert`, I guess the current approach is probably the safest, although it has the drawback of needing to keep the JSON file around as well. Another future possibility would be to accept `languages.json` file as well

Reply:
I wanted to avoid calling Nickel because I was afraid of `import-from-derivation`, but in fact it was no problem in `fromNickelFile`. I can't really judge which solution is best, but I'll happily follow your decision on this!